File: figs/geohash_grids.lua
package.path = package.path .. ';lib/?.lua'
local luamp = require 'luamp'
local stream = require 'stream'
local table = require 'table'

local figs = {}

table.insert(
    figs,
    luamp.polygon({
        luamp.point(0.8, 0.7),
        luamp.point(0.5, 1.4),
        luamp.point(1.1, 2.8),
        luamp.point(2.2, 2.5),
        luamp.point(2.5, 1.1),
        luamp.point(1.2, 0.5)},
        {brush_color=luamp.colors.gray, pen_color=luamp.colors.invisible}))
stream
    .iterate(
        luamp.origin,
        function(x)
            return x + luamp.point(1, 0)
        end)
    :map(function(x)
        return luamp.line(
            x,
            x + luamp.point(0, 3),
            {line_style=luamp.line_styles.dashed})
    end)
    :take(4)
    :collect(figs)
stream
    .iterate(
        luamp.origin,
        function(x)
            return x + luamp.point(0, 1)
        end)
    :map(function(x)
        return luamp.line(
            x,
            x + luamp.point(3, 0),
            {line_style=luamp.line_styles.dashed})
    end)
    :take(4)
    :collect(figs)

print(luamp.figure(table.unpack(figs)))

File: figs/geohash_grids.py
from fathom import Point, ORIGIN
from fathom.tikz import Canvas
from fathom.geometry import *
import fathom.tikz.colors as colors
import fathom.tikz.line_styles as line_styles
import fathom.tikz.locations as locations
from itertools import *

if __name__ == '__main__':
    canvas = Canvas(
        preamble=['\\usepackage{amsmath}'],
        leading_instructions=['\\footnotesize'])

    canvas.new_polygon(
        vertices=[
            Point(0.8, 0.7),
            Point(0.5, 1.4),
            Point(1.1, 2.8),
            Point(2.2, 2.5),
            Point(2.5, 1.1),
            Point(1.2, 0.5)],
        brush_color=colors.GRAY,
        pen_color=colors.INVISIBLE)

    vs = repeat(ORIGIN)
    vs = accumulate(vs, lambda x, _: x + Point(1, 0))
    vs = list(islice(vs, 4))
    for x in vs:
        canvas.new_line(
            src=x,
            dst=x + Point(0, 3),
            line_style=line_styles.DASHED)

    vs = repeat(ORIGIN)
    vs = accumulate(vs, lambda x, _: x + Point(0, 1))
    vs = list(islice(vs, 4))
    for x in vs:
        canvas.new_line(
            src=x,
            dst=x + Point(3, 0),
            line_style=line_styles.DASHED)

    print(canvas.draw())
